File: VB/GridViewFixedRows/GridViewFixedRows.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.Utils.Serializing
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Drawing

Namespace GVFixedRows
	Friend Class GridViewFixedRows
		Inherits GridView
		Public Sub New(ByVal grid As DevExpress.XtraGrid.GridControl)
			MyBase.New(grid)
		fixedRows_Renamed = New List(Of DataRow)()
		End Sub

		Public Sub New()
			Me.New(Nothing)
		AddHandler CustomRowFilter, AddressOf GridViewFixedRow_CustomRowFilter
		End Sub

		Private Sub GridViewFixedRow_CustomRowFilter(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.RowFilterEventArgs)
			Dim dataRow As DataRow = (TryCast(Me.DataSource, DataView)).Table.Rows(e.ListSourceRow)
			If Me.IsFixedRow(dataRow) Then
				e.Visible = False
				e.Handled = True
			End If
		End Sub

		Protected Overrides ReadOnly Property ViewName() As String
			Get
				Return "GridViewFixedRows"
			End Get
		End Property

		Private fixedRows_Renamed As List(Of DataRow)
		Public Overridable Sub AddToFixedRows(ByVal dataRow As DataRow)
			If (Not fixedRows_Renamed.Contains(dataRow)) Then
				fixedRows_Renamed.Add(dataRow)
				RefreshData()
			End If
		End Sub

		Public Overridable Sub RemoveFromFixedRows(ByVal dataRow As DataRow)
			fixedRows_Renamed.Remove(dataRow)
			RefreshData()
		End Sub

		Public Overridable Function IsFixedRow(ByVal dataRow As DataRow) As Boolean
			Return fixedRows_Renamed.Contains(dataRow)
		End Function

		Public Overridable ReadOnly Property FixedRows() As List(Of DataRow)
			Get
				Return fixedRows_Renamed
			End Get
		End Property

		Public Function InFixedRowsPane(ByVal point As Point) As Boolean
			Dim rects As GridViewRectsFixedRows = TryCast(Me.ViewInfo.ViewRects, GridViewRectsFixedRows)
			If GridDrawing.PtInRect(rects.FixedRows, point) Then
				Return True
			Else
				Return False
			End If
		End Function

		Public Function GetFixedDataRow(ByVal point As Point) As DataRow
			Dim rects As GridViewRectsFixedRows = TryCast(Me.ViewInfo.ViewRects, GridViewRectsFixedRows)
			Dim Y As Integer = rects.FixedRows.Y
			For i As Integer = 0 To Me.FixedRows.Count - 1
				If point.Y >= Y AndAlso point.Y <= Y + Me.ViewInfo.ColumnRowHeight Then
					Return Me.FixedRows(i)
				End If
				Y += Me.ViewInfo.ColumnRowHeight
			Next i
			Return Nothing
		End Function
	End Class
End Namespace

File: CS/GridViewFixedRows/GridViewFixedRows.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.Utils.Serializing;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Drawing;

namespace GVFixedRows
{
    class GridViewFixedRows : GridView
    {
        public GridViewFixedRows(DevExpress.XtraGrid.GridControl grid) : base(grid) 
        { fixedRows = new List<DataRow>(); }

        public GridViewFixedRows() : this(null)
        { this.CustomRowFilter += new DevExpress.XtraGrid.Views.Base.RowFilterEventHandler(GridViewFixedRow_CustomRowFilter); }

        void GridViewFixedRow_CustomRowFilter(object sender, DevExpress.XtraGrid.Views.Base.RowFilterEventArgs e)
        {
            DataRow dataRow = (this.DataSource as DataView).Table.Rows[e.ListSourceRow];
            if (this.IsFixedRow(dataRow))
            {
                e.Visible = false;
                e.Handled = true;
            }
        }

        protected override string ViewName { get { return "GridViewFixedRows"; } }

        List<DataRow> fixedRows;
        public virtual void AddToFixedRows(DataRow dataRow)
        {
            if (!fixedRows.Contains(dataRow))
            {
                fixedRows.Add(dataRow);
                RefreshData();
            }
        }

        public virtual void RemoveFromFixedRows(DataRow dataRow)
        {
            fixedRows.Remove(dataRow);
            RefreshData();
        }

        public virtual bool IsFixedRow(DataRow dataRow)
        { return fixedRows.Contains(dataRow); }

        public virtual List<DataRow> FixedRows { get { return fixedRows; } }

        public bool InFixedRowsPane(Point point)
        {
            GridViewRectsFixedRows rects = this.ViewInfo.ViewRects as GridViewRectsFixedRows;
            if (GridDrawing.PtInRect(rects.FixedRows, point))
                return true;
            else
                return false;
        }

        public DataRow GetFixedDataRow(Point point)
        {
            GridViewRectsFixedRows rects = this.ViewInfo.ViewRects as GridViewRectsFixedRows;
            int Y = rects.FixedRows.Y;
            for (int i = 0; i < this.FixedRows.Count; i++)
            {
                if (point.Y >= Y && point.Y <= Y + this.ViewInfo.ColumnRowHeight)
                    return this.FixedRows[i];
                Y += this.ViewInfo.ColumnRowHeight;
            }
            return null;
        }
    }
}
